sql,provisioning: enable provisioning during LDAP authentication

souravcrl opened the pull request:
After successful authentication with LDAP, CRDB will now also provision the LDAP
user corresponding to the db user from the connection string. The user needs to
have valid credentials for binding to LDAP server and we will skip other
validations for pre-existence of the user and privilege to perform cluster sql
login.

informs #147602
fixes #147599
Epic CRDB-21590

Release note (enterprise change): Added a new cluster setting
`server.provisioning.ldap.enabled` which can be set to true to conditionally
enable user provisioning during sql cluster authentication. The user
authenticates with the LDAP server and CRDB will only validate that bind to the
IDP was successful for provisioning the user. All roles created thus will be
privileged to perform sql authentication and will mandatorily have a role option
for PROVISIONING_SOURCE set to `ldap:<idp_url>`. Any group roles that are to be
assigned via ldap authorization must be pre created prior to the authentication
start.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/148200)
<!-- Reviewable:end -->

pritesh-lahoti commented on pkg/security/provisioning/settings.go:
```diff
@@ -0,0 +1,67 @@
+const (
+	supportedAuthMethodLDAP           = "ldap"
+	baseProvisioningSettingName       = "security.provisioning."
+	ldapProvisioningEnableSettingName = baseProvisioningSettingName + "ldap.enabled"
```
As per the design, are we not namespacing the cluster setting under `server.ldap_authentication.`?

souravcrl replied:
I actually wanted to scope it separately as the provisioning package exists outside of the jwt/ldap modules and affects authentication, authorization as well other auth behavior like user role remapping, etc. Hence wanted to scope it separately under `security.provisioning` and then `ldap.enabled`/`jwt.enabled` for each access method. LMK if this should change as I had earlier not factored in role_remapping/authZ into my decision for naming the setting.
<!-- Reviewable comment -OSu-wDo7G7LNEUAh4Q8:b-ygbe3o -->
<!-- Sent from Reviewable.io -->

pritesh-lahoti commented on pkg/security/provisioning/settings.go:
Do we need to have this type now? Do we plan to use it in the near future?

souravcrl replied:
This is to keep parity with base non-implemented type for implementable interfaces, for e.g. we have: https://github.com/cockroachdb/cockroach/blob/fea6fe50e0ebc842315a96ab4247df0f60784cf8/pkg/sql/pgwire/auth_methods.go#L739-L762 for `jwtVerifierConfig`.
LMK if we should rename it to `noProvisioningConfig` to make it similar to that but felt this seemed more intuitive as feature doesn't affect cli.  
<!-- Reviewable comment -OSu-yHj2JBZK5mrDcA1:breyp8e -->
<!-- Sent from Reviewable.io -->

souravcrl replied:
updated it to be `noProvisioningConfig` as per discussion.

souravcrl reviewed (commented):
I have added some negative cases to the ldap data driven tests.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-:-OSvwTq42x5FshGr7XEU:b-10hcsw)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pritesh-lahoti and @rafiss)



<!-- Sent from Reviewable.io -->

spilchen reviewed (changes requested):
Looks good. Nothing major. Just mainly had questions.

> Reviewed 7 of 7 files at r1, 23 of 23 files at r2, 8 of 8 files at r3, 13 of 13 files at r5, 13 of 13 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-:-OTCMA3D5_C_kdUVpZdX:bt6cx2s)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @alyshanjahani-crl, @arjunmahishi, @pritesh-lahoti, and @rafiss)

___
*[`pkg/security/provisioning/settings.go` line 27 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTC5XCiDAzU-ItTDH3S:-OTC5XCiDAzU-ItTDH3T:bi5b6l7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/security/provisioning/settings.go#L27)):*
> ```Go
> 	settings.ApplicationLevel,
> 	ldapProvisioningEnableSettingName,
> 	"enables or disables automatic user provisioning for ldap authentication method",
> ```

Seeing as this ends up being document since it's an external var, I suggest slightly different wording of the description. Provisioning is in the setting name, maybe describe it without using that term to make it clear. Something like this:
"enables automatic creation of SQL users upon successful LDAP login"
___
*[`pkg/security/provisioning/settings.go` line 57 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTCI7u1ELUTymTYllkl:-OTCI7u1ELUTymTYllkm:b-ind1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/security/provisioning/settings.go#L57)):*
> ```Go
> 
> // noProvisioningConfig defines the base user provisioning settings.
> type noProvisioningConfig struct{}
> ```

Do we use this config yet? I couldn't find where we use it.
___
*[`pkg/sql/pgwire/auth.go` line 166 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTCJlxB-T7GtLxuqOr9:-OTCJlxB-T7GtLxuqOrA:blqtjru) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/sql/pgwire/auth.go#L166)):*
> ```Go
> 	}
> 
> 	if behaviors.IsProvisioningEnabled(execCfg.Settings, hbaEntry.Method.String()) {
> ```

Do we need a version gate in here somewhere? Or is it fine to come in here if we haven't yet finalized on 25.3 yet.
___
*[`pkg/sql/pgwire/auth_methods.go` line 1008 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTCFDAA5J6SbbN4I3ZH:-OTCFDAA5J6SbbN4I3ZI:b6nunf9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/sql/pgwire/auth_methods.go#L1008)):*
> ```Go
> 
> 	b.SetProvisioner(func(ctx context.Context) error {
> 		c.LogAuthInfof(ctx, "LDAP authentication succeeded; attempting to provision user")
> ```

Do we want to log something on the exit of this closure to indicate the provisioning was successful?
___
*[`pkg/sql/create_role.go` line 365 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTCGGs4Brmclf7fA7vB:-OTCGGs54MkvX3DLsCq1:b-9dz0lz) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/sql/create_role.go#L365)):*
> ```Go
> ) error {
> 	return execCfg.InternalDB.DescsTxn(ctx, func(ctx context.Context, txn descs.Txn) error {
> 		if len(provisioningSource) > 0 {
> ```

Why is it not considered an error if the provisioningSource is empty? Or, what cases would this be valid? It seems like we will silently skip creating the user in that case. That could introduce confusion. 
___
*[`pkg/sql/create_role.go` line 366 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTCAIWJ9uA_amPuVcSD:-OTCAIWJ9uA_amPuVcSE:b-xw94j2) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/sql/create_role.go#L366)):*
> ```Go
> 	return execCfg.InternalDB.DescsTxn(ctx, func(ctx context.Context, txn descs.Txn) error {
> 		if len(provisioningSource) > 0 {
> 			userProvisioningStmt := fmt.Sprintf("CREATE USER IF NOT EXISTS %s WITH PROVISIONING_SOURCE %q", user.SQLIdentifier(), provisioningSource)
> ```

Are there users that we don't want to do this for? Possibly `user.IsRootUser()` or if `user.IsReserved()` is true. If yes, are these errors or no-ops?


<!-- Sent from Reviewable.io -->

souravcrl committed:
sql,provisioning: enable provisioning during LDAP authentication

After successful authentication with LDAP, CRDB will now also provision the LDAP
user corresponding to the db user from the connection string. The user needs to
have valid credentials for binding to LDAP server and we will skip other
validations for pre-existence of the user and privilege to perform cluster sql
login.

informs #147602
fixes #147599
Epic CRDB-21590

Release note (enterprise change): Added a new cluster setting
`server.provisioning.ldap.enabled` which can be set to true to conditionally
enable user provisioning during sql cluster authentication. The user
authenticates with the LDAP server and CRDB will only validate identity lookup
on IDP was successful for provisioning the user. All roles created thus will be
privileged to perform sql authentication and will mandatory have a role option
for PROVISIONSRC set to `ldap:<idp_url>`. Any group roles that are to be
assigned via ldap authorization must be pre created prior to the authentication
start.

souravcrl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148200)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @alyshanjahani-crl, @arjunmahishi, @pritesh-lahoti, @rafiss, and @spilchen)

___
*[`pkg/security/provisioning/settings.go` line 27 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTC5XCiDAzU-ItTDH3S:-OTlEkSzA9HP8SXR64r8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/security/provisioning/settings.go#L27)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Seeing as this ends up being document since it's an external var, I suggest slightly different wording of the description. Provisioning is in the setting name, maybe describe it without using that term to make it clear. Something like this:
"enables automatic creation of SQL users upon successful LDAP login"
</blockquote></details>

Done.
___
*[`pkg/security/provisioning/settings.go` line 57 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTCI7u1ELUTymTYllkl:-OTlEl4h69uaYTDEfSnp:bfeamc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/security/provisioning/settings.go#L57)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Do we use this config yet? I couldn't find where we use it.
</blockquote></details>

This is used as a compile time assertion for the type `UserProvisioningConfig` to prevent api drift and safety refactoring.
___
*[`pkg/sql/create_role.go` line 365 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTCGGs4Brmclf7fA7vB:-OTlHl5qDr_xpSkfNymy:b3g8qec) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/sql/create_role.go#L365)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Why is it not considered an error if the provisioningSource is empty? Or, what cases would this be valid? It seems like we will silently skip creating the user in that case. That could introduce confusion.
</blockquote></details>

I realized we don't need this validation as we already parse the role option value during assignment. Hence removing the check.
___
*[`pkg/sql/create_role.go` line 366 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTCAIWJ9uA_amPuVcSD:-OTlIXBZ9qi91dhHtTP_:bmv58yu) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/sql/create_role.go#L366)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Are there users that we don't want to do this for? Possibly `user.IsRootUser()` or if `user.IsReserved()` is true. If yes, are these errors or no-ops?
</blockquote></details>

I feel all the roles except root and reserved users should be provisionable. For the above 2 cases, isn't it expected that the sql command `CREATE USER` will fail and the generated error could be  the expected output here?
___
*[`pkg/sql/pgwire/auth.go` line 166 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTCJlxB-T7GtLxuqOr9:-OTlKbhh6do3PYSXMbXz:b69eq41) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/sql/pgwire/auth.go#L166)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Do we need a version gate in here somewhere? Or is it fine to come in here if we haven't yet finalized on 25.3 yet.
</blockquote></details>

I am unsure if this will be required since we are gating the role creation here https://github.com/cockroachdb/cockroach/blob/7f725b06b37f0b2caa6dc0496404ba21072e586a/pkg/sql/roleoption/role_option.go#L238-L251 for 25_3. Even if the feature is enabled since the usability of the role option is only post 25_3, the feature is unusable.
___
*[`pkg/sql/pgwire/auth_methods.go` line 1008 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTCFDAA5J6SbbN4I3ZH:-OTlNfC92zlEV2T_-zoD:b-3f0wrj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/sql/pgwire/auth_methods.go#L1008)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Do we want to log something on the exit of this closure to indicate the provisioning was successful?
</blockquote></details>

I feel this won't add any additional observability here as we log before the starting of the auth module like `attempting to provision user` or `attempting authorization`. Any errors are still logged, so we don't miss out on events. LMK if there is flaw in this model though as there maybe some cases I am missing.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> Reviewed 11 of 46 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148200)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @alyshanjahani-crl, @arjunmahishi, @pritesh-lahoti, @rafiss, and @souravcrl)

___
*[`pkg/security/provisioning/settings.go` line 57 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTCI7u1ELUTymTYllkl:-OTmYBjM-RV5ow-UICS8:b-jrfmcf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/security/provisioning/settings.go#L57)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

This is used as a compile time assertion for the type `UserProvisioningConfig` to prevent api drift and safety refactoring.
</blockquote></details>

I understand the purpose of `var _ UserProvisioningConfig = ...` as a compile-time check. It definitely makes sense for concrete structs like `clusterProvisioningConfig` that are actually used. But does it still provide value for `noProvisioningConfig`, which isn’t used anywhere? Since we already have the check for `clusterProvisioningConfig`, we're covered in terms of interface drift. Just wondering if the extra assertion is needed in this case.
___
*[`pkg/sql/create_role.go` line 366 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTCAIWJ9uA_amPuVcSD:-OTmZda_7iGNrXE_UOho:b-bxy4ux) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/sql/create_role.go#L366)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

I feel all the roles except root and reserved users should be provisionable. For the above 2 cases, isn't it expected that the sql command `CREATE USER` will fail and the generated error could be  the expected output here?
</blockquote></details>

ah yes, good point. You can disregard this comment.
___
*[`pkg/sql/pgwire/auth.go` line 166 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTCJlxB-T7GtLxuqOr9:-OTmb0rW7W8xVvM_O9hu:bskgnkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/sql/pgwire/auth.go#L166)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

I am unsure if this will be required since we are gating the role creation here https://github.com/cockroachdb/cockroach/blob/7f725b06b37f0b2caa6dc0496404ba21072e586a/pkg/sql/roleoption/role_option.go#L238-L251 for 25_3. Even if the feature is enabled since the usability of the role option is only post 25_3, the feature is unusable.
</blockquote></details>

It might be okay. I just want to fully understand this. The scenario is:
1. have HBA config setup for LDAP authentication
2. move to a 25.3 build, but don't finalize the cluster
3. set `server.ldap_authentication.provisioning.enabled` to true
4. login

At step 4, if the user logs in with a user that already exists in CRDB, there are no issues. But if they login with a user that exists in LDAP but not in CRDB, then it will enter the auto-provisioning code. When it attempts to run `CREATE USER` with the `PROVISIONSRC` role option, the create will fail. And the entire login process will fail with: PROVISIONSRC role option is only supported after v25.3 upgrade is finalized.

Is that correct?

If so, a few follow-up questions:
- Is there a mixed-version test to verify this? If not, can we create one?
- Does the error make sense? Will it be obvious to the end user what PROVISIONSRC is and why it was being added? There's a clue that it dependent on 25.3 upgrade, and maybe that is okay.
___
*[`pkg/sql/pgwire/auth_methods.go` line 1008 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTCFDAA5J6SbbN4I3ZH:-OTmdXgR1pujOXXdak8h:b-dxcbts) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/sql/pgwire/auth_methods.go#L1008)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

I feel this won't add any additional observability here as we log before the starting of the auth module like `attempting to provision user` or `attempting authorization`. Any errors are still logged, so we don't miss out on events. LMK if there is flaw in this model though as there maybe some cases I am missing.
</blockquote></details>

Makes sense.


<!-- Sent from Reviewable.io -->

souravcrl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148200)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @alyshanjahani-crl, @arjunmahishi, @pritesh-lahoti, @rafiss, and @spilchen)

___
*[`pkg/security/provisioning/settings.go` line 57 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTCI7u1ELUTymTYllkl:-OU0apAN6YuiRjV5DVIg:b-dtjnpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/security/provisioning/settings.go#L57)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I understand the purpose of `var _ UserProvisioningConfig = ...` as a compile-time check. It definitely makes sense for concrete structs like `clusterProvisioningConfig` that are actually used. But does it still provide value for `noProvisioningConfig`, which isn’t used anywhere? Since we already have the check for `clusterProvisioningConfig`, we're covered in terms of interface drift. Just wondering if the extra assertion is needed in this case.
</blockquote></details>

I have removed the `noProvisioningConfig` inline with the discussion here. I was unsure if a no-op assertion is always required or not but this makes it clear.
___
*[`pkg/sql/pgwire/auth.go` line 166 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OTCJlxB-T7GtLxuqOr9:-OU0b6Ro49_LgxqlsJlR:bcbjlmo) ([raw file](https://github.com/cockroachdb/cockroach/blob/2637aa4e459ac4126ea63f353d7bef6e1d905fe5/pkg/sql/pgwire/auth.go#L166)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

It might be okay. I just want to fully understand this. The scenario is:
1. have HBA config setup for LDAP authentication
2. move to a 25.3 build, but don't finalize the cluster
3. set `server.ldap_authentication.provisioning.enabled` to true
4. login

At step 4, if the user logs in with a user that already exists in CRDB, there are no issues. But if they login with a user that exists in LDAP but not in CRDB, then it will enter the auto-provisioning code. When it attempts to run `CREATE USER` with the `PROVISIONSRC` role option, the create will fail. And the entire login process will fail with: PROVISIONSRC role option is only supported after v25.3 upgrade is finalized.

Is that correct?

If so, a few follow-up questions:
- Is there a mixed-version test to verify this? If not, can we create one?
- Does the error make sense? Will it be obvious to the end user what PROVISIONSRC is and why it was being added? There's a clue that it dependent on 25.3 upgrade, and maybe that is okay.
</blockquote></details>

I have the added the mixed version tests. Currently the register for ldap in HBA config exists outside of the pgwire package, hence it is not available when logictests are run and logictests only test for the `default` cert-password auth method. These tests do not support for HBA logic as well as we test them separately in `pgwire/auth_test.go`. I have added some test specific conditions to the `provisioning.Enabled(authmethod)` so it can be properly tested for the mixed version logic tests.


<!-- Sent from Reviewable.io -->

spilchen reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Just a minor question.

> Reviewed 8 of 8 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-:-OU2N3qT9PIa1s1wD8XA:b3jemk9)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @alyshanjahani-crl, @arjunmahishi, @pritesh-lahoti, and @rafiss)

___
*[`pkg/sql/pgwire/auth_behaviors.go` line 173 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OU2MIAX1DIKcopjBuLA:-OU2MIAX1DIKcopjBuLB:b-adzfyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad3f6cfa4b3dce6475b6d3f7c2fbca1d43d46486/pkg/sql/pgwire/auth_behaviors.go#L173)):*
> ```Go
> 				return found(ctx)
> 			} else {
> 				return errors.New("no provisioner provided to AuthBehaviors")
> ```

Should this error, and the one below it, have a pgcode attached to it (i.e. use `pgerror.Newf`)? Or is the error wrapped later with an appropriate code?


<!-- Sent from Reviewable.io -->

souravcrl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148200)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @alyshanjahani-crl, @arjunmahishi, @pritesh-lahoti, @rafiss, and @spilchen)

___
*[`pkg/sql/pgwire/auth_behaviors.go` line 173 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/148200#-OU2MIAX1DIKcopjBuLA:-OU3kBUB6gRlxrypRWnq:b-7sezfa) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad3f6cfa4b3dce6475b6d3f7c2fbca1d43d46486/pkg/sql/pgwire/auth_behaviors.go#L173)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Should this error, and the one below it, have a pgcode attached to it (i.e. use `pgerror.Newf`)? Or is the error wrapped later with an appropriate code?
</blockquote></details>

I am currently the `pgcode.InvalidAuthorizationSpecification` upstream in https://github.com/cockroachdb/cockroach/blob/ad3f6cfa4b3dce6475b6d3f7c2fbca1d43d46486/pkg/sql/pgwire/auth.go#L174 Should this change because provisioning is exclusive of authorization but I did not find any other appropriate code for this?


<!-- Sent from Reviewable.io -->

souravcrl commented:
TFTR!
bors=pritesh-lahoti,spilchen

souravcrl commented:
bors r=pritesh-lahoti,spilchen
